core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -1413,55 +1413,106 @@ final class Stream[+F[_], +O] private[fs2] (private[fs2] val underlying: Pull[F,
-            : Pull[F2, Chunk[O], Unit] =
-          timedPull.uncons.flatMap {
-            case None =>
-              Pull.output1(acc).whenA(acc.nonEmpty)
-            case Some((e, next)) =>
-              def resetTimerAndGo(q: Chunk[O]) =
-                timedPull.timeout(timeout) >> go(q, next)
-
-              e match {
-                case Left(_) =>
-                  if (acc.nonEmpty)
-                    Pull.output1(acc) >> resetTimerAndGo(Chunk.empty)
-                  else
-                    go(Chunk.empty, next, hasTimedOut = true)
-                case Right(c) if hasTimedOut =>
-                  // it has timed out without reset, so acc is empty
-                  val (toEmit, rest) =
-                    if (c.size < n) Pull.output1(c) -> Chunk.empty
-                    else resize(c, Pull.done)
-                  toEmit >> resetTimerAndGo(rest)
-                case Right(c) =>
-                  val newAcc = acc ++ c
-                  if (newAcc.size < n)
-                    go(newAcc, next)
-                  else {
-                    val (toEmit, rest) = resize(newAcc, Pull.done)
-                    toEmit >> resetTimerAndGo(rest)
+        def dequeueN(n: Int): F2[Option[Vector[O]]] =
+          acquireSupplyUpToNWithin(n.toLong, timeout).flatMap { n =>
+            buffer
+              .modify { buf =>
+                if (buf.data.size >= n) {
+                  val (head, tail) = buf.data.splitAt(n.toInt)
+                  (buf.copy(tail), buf.copy(head))
+                } else {
+                  (buf.copy(Vector.empty), buf)
+                }
+              }
+              .flatMap { buf =>
+                demand.releaseN(buf.data.size.toLong).flatMap { _ =>
+                  buf.endOfSupply match {
+                    case Some(Left(error)) =>
+                      F.raiseError(error)
+                    case Some(Right(_)) if buf.data.isEmpty =>
+                      F.pure(None)
+                    case _ =>
+                      F.pure(Some(buf.data))
                   }
+                }
               }
           }
 
-        timedPull.timeout(timeout) >> go(Chunk.empty, timedPull)
+        def endSupply(result: Either[Throwable, Unit]): F2[Unit] =
+          buffer.update(_.copy(endOfSupply = Some(result))) *> supply.releaseN(Int.MaxValue)
+
+        def endDemand(result: Either[Throwable, Unit]): F2[Unit] =
+          buffer.update(_.copy(endOfDemand = Some(result))) *> demand.releaseN(Int.MaxValue)
+
+        val enqueueAsync = F.start {
+          this
+            .evalMap(enqueue)
+            .forall(identity)
+            .onFinalizeCase {
+              case ExitCase.Succeeded  => endSupply(Right(()))
+              case ExitCase.Errored(e) => endSupply(Left(e))
+              case ExitCase.Canceled   => endSupply(Right(()))
+            }
+            .compile
+            .drain
+        }
+
+        fs2.Stream
+          .eval(enqueueAsync)
```

Comment:
if this gets cancelled, you won't interrupt the producer stream

Reply:
Won't that work: `endDemand(ending) *> upstream.cancel`?

Reply:
if the stream gets canceled during `eval`, you won't reach that point. The `concurrently` combinator covers this general use case for you. (there is a general meta point here about perf vs correctness, I'm finding it fairly hard to fully reason about the correctness of this implementation, even though I believe it's important enough to justify high performance)

Reply:
perhaps you can move the `onFinalize` one level up though

Reply:
`bracketCase` should do the trick I think